src/api/utils/tradeQuoter.ts

```diff
@@ -0,0 +1,451 @@
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+'use strict';
+
+import BigDecimal from 'js-big-decimal';
+import { BigNumber, FixedNumber, utils as ethersUtils } from 'ethers';
+import type TradeModuleWrapper from '@src/wrappers/set-protocol-v2/TradeModuleWrapper';
+
+import {
+  CoinGeckoCoinPrices,
+  QuoteOptions,
+  TradeQuote,
+} from '../../types/index';
+
+import {
+  CoinGeckoDataService,
+  USD_CURRENCY_CODE
+} from './coingecko';
+
+import { Address } from '@setprotocol/set-protocol-v2/utils/types';
+import { GasOracleService } from './gasOracle';
+import { ZeroExTradeQuoter } from './zeroex';
+
+export const ZERO_EX_ADAPTER_NAME = 'ZeroExApiAdapterV3';
+
+const SCALE = BigNumber.from(10).pow(18);
+
+/**
+ * @title TradeQuote
+ * @author Set Protocol
+ *
+ * A utility library to generate trade quotes for token pairs associated with a
+ * set for Ethereum and Polygon chains. Uses 0xAPI to get quote and requires a valid
+ * 0x api key.
+ */
+
+export class TradeQuoter {
+  private tradeQuoteGasBuffer: number = 5;
+  private feeRecipient: Address = '0xD3D555Bb655AcBA9452bfC6D7cEa8cC7b3628C55';
+  private feePercentage: number = 0;
+  private isFirmQuote: boolean = true;
+  private slippagePercentage: number = 2;
+  private excludedSources: string[] = ['Kyber', 'Eth2Dai', 'Uniswap', 'Mesh'];
+  private zeroExApiKey: string;
+
+  constructor(zeroExApiKey: string = '') {
+    this.zeroExApiKey = zeroExApiKey;
+  }
+
+  /**
+   * Generates a trade quote for a token pair in a SetToken. This method requires
+   * a token metadata map (passed with the options) which can be generated using
+   * the CoinGeckoDataService in '.api/utils/coingecko.ts'.
+   *
+   * @param  options  QuoteOptions: options / config to generate the quote
+   * @return          TradeQuote: trade quote object
+   */
+  async generate(options: QuoteOptions): Promise<TradeQuote> {
+    const chainId = options.chainId;
+    const feePercentage = options.feePercentage || this.feePercentage;
+    const isFirmQuote = (options.isFirmQuote === false) ? false : this.isFirmQuote;
+    const slippagePercentage = options.slippagePercentage || this.slippagePercentage;
+    const feeRecipient = options.feeRecipient || this.feeRecipient;
+    const excludedSources = options.excludedSources || this.excludedSources;
+
+    const exchangeAdapterName = ZERO_EX_ADAPTER_NAME;
+
+    const {
+      fromTokenAddress,
+      toTokenAddress,
+      fromAddress,
+    } = this.sanitizeAddress(options.fromToken, options.toToken, options.fromAddress);
+
+    const amount = this.sanitizeAmount(options.rawAmount, options.fromTokenDecimals);
+
+    const setOnChainDetails = await options.setToken.fetchSetDetailsAsync(
```

Comment:
I forgot we also make this call. Feels redundant since the client will have all of this information as well as the manager address.. The data only serves a check that the component is a component in the Set.

The manager address is going to be required to fetch a gas estimate since only the manager can submit the trade.